src/avram/bulk_upsert.cr

```diff
@@ -0,0 +1,70 @@
+class Avram::BulkUpsert(T)
+  def initialize(@records : Array(T), @column_names : Array(Symbol))
+    @records = set_timestamps(records)
+  end
+
+  def statement
+    [
+      "INSERT INTO #{table}(#{fields})",
+      "VALUES #{value_placeholders}",
+      "ON CONFLICT (#{conflicts}) DO UPDATE SET #{updates}",
+      "RETURNING #{returning}",
+    ].join(" ")
+  end
+
+  private def conflicts
+    @column_names.join(", ")
+  end
+
+  private def set_timestamps(collection)
+    collection.map do |record|
+      record.created_at.value ||= Time.utc if record.responds_to?(:created_at)
+      record.updated_at.value = Time.utc if record.responds_to?(:updated_at)
```

Comment:
Yeah we were considering this! The one thing we needed to work through for this was a MAJOR data changeup -- these would then need to be NOW() instead of bound $n params which changes up the models a lot

Reply:
Related: https://github.com/luckyframework/avram/issues/393 (note the issue on `updated_at` here that would require a trigger to know when to change)
Somewhat related: https://github.com/luckyframework/lucky/discussions/1583